### PR TITLE
Make a common standard API for querying ufunc impl

### PR DESCRIPTION
**Pull request by sklam**
Expanding on https://github.com/numba/numba/pull/6467

- Instead of directly exposing `ufunc_db`, expose a `Context.get_ufunc_info(ufunc_key)` which mirrors the functionality of `ufunc_db.get_ufunc_info`.
- no longer need to branch in npyimpl.py.

### CODE REVIEW

File: numba/core/base.py [hunk 1]
```diff
@@ -1146,6 +1146,24 @@ def add_linking_libs(self, libs):
         for lib in libs:
             colib.add_linking_library(lib)
 
+    def get_ufunc_info(self, ufunc_key):
+        """Get ufunc implementation given a ufunc object as the key.
```
Comment (stuartarchibald) [inline numba/core/base.py:1150]:
```suggestion
        """Get the ufunc implementation for a given ufunc object.
```

File: numba/core/base.py [hunk 2]
```diff
@@ -1146,6 +1146,24 @@ def add_linking_libs(self, libs):
...
+
+        The default implementation in BaseContext always raise a
```
Comment (stuartarchibald) [inline numba/core/base.py:1152]:
```suggestion
        The default implementation in BaseContext always raises a
```

File: numba/core/base.py [hunk 3]
```diff
@@ -1146,6 +1146,24 @@ def add_linking_libs(self, libs):
...
+        ``NotImplementedError`` exception. Subclasses may raise ``KeyError``
+        to signal that the given ``ufunc_key`` is not available.
+
+        Parameters
+        ----------
+        ufunc_key : numpy ufunc
```
Comment (stuartarchibald) [inline numba/core/base.py:1158]:
```suggestion
        ufunc_key : NumPy ufunc
```

File: numba/core/base.py [hunk 4]
```diff
@@ -1146,6 +1146,24 @@ def add_linking_libs(self, libs):
...
+
+        Returns
+        -------
+        res : dict[str, callable]
+            A mapping of NumPy ufunc type signature to the lower-level
```
Comment (stuartarchibald) [inline numba/core/base.py:1163]:
```suggestion
            A mapping of a NumPy ufunc type signature to a lower-level
```

File: numba/np/npyimpl.py [hunk 5]
```diff
@@ -539,7 +539,7 @@ def generate(self, *args):
 
 def _register_ufuncs():
     kernels = {}
-
+    # Assuming ufunc implementation for the CPUContext.
```
Comment (stuartarchibald) [inline numba/np/npyimpl.py:542]:
```suggestion
    # NOTE: Assuming ufunc implementation for the CPUContext.
```

File: numba/np/numpy_support.py [hunk 6]
```diff
@@ -279,6 +279,7 @@ def supported_ufunc_loop(ufunc, loop):
     legacy and when implementing new ufuncs the ufunc_db should be preferred,
     as it allows for a more fine-grained incremental support.
     """
+    # Assuming ufunc for the CPUContext
```
Comment (stuartarchibald) [inline numba/np/numpy_support.py:282]:
```suggestion
    # NOTE: Assuming ufunc for the CPUContext
```

File: numba/tests/test_ufuncs.py [hunk 7]
```diff
@@ -1768,5 +1771,33 @@ def wrapper():
...
+        expected = ufunc_db.get_ufunc_info(np.add)
+        self.assertEqual(add_info, expected)
+        # Check: KeyError raised on bad key
+        badkey = object()
+        with self.assertRaises(KeyError) as raises:
+            ufunc_db.get_ufunc_info(badkey)
+        self.assertEqual(raises.exception.args, (badkey,))
+
+    def test_base_get_ufunc_info(self):
+        # The BaseContext always raise NotImplementedError
```
Comment (stuartarchibald) [inline numba/tests/test_ufuncs.py:1792]:
```suggestion
        # The BaseContext always raises NotImplementedError
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, just a few minor typos else looks good.

**Comment by sklam** [comment]
@reazulhoque, @ PokhodenkoSA, This is my proposed changes to https://github.com/numba/numba/pull/6467 to convert the dynamically checked `ufunc_db` part of the target context interface.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes, looks good.

**Comment by reazulhoque** [comment]
@sklam The changes look good.

**Comment by stuartarchibald** [comment]
> @sklam The changes look good.

Thanks for checking.

**Pull request merged**